server/src/database/MainDatabase.hh

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include <oatpp/orm/DbClient.hpp>
+#include <oatpp/core/macro/codegen.hpp>
+#include <oatpp/orm/SchemaMigration.hpp>
+#include <filesystem>
+#include <ranges>
+#include <vector>
+
+#include "dto/UserDto.hh"
+#include "dto/PlaceDto.hh"
+#include "dto/OfferDto.hh"
+
+#include OATPP_CODEGEN_BEGIN(DbClient)
+
+namespace server::database {
+
+    class MainDatabase : public oatpp::orm::DbClient {
+    public:
+        [[nodiscard]]
+        explicit
+        MainDatabase(std::shared_ptr<oatpp::orm::Executor> const &executor)
+                : oatpp::orm::DbClient{executor} {
+            oatpp::orm::SchemaMigration migrator{executor};
+            auto migrations =
+                    std::filesystem::directory_iterator{MAIN_DATABASE_MIGRATIONS_PATH}
+                    | std::views::all | std::views::filter([](auto const &entry) {
+                        return entry.is_regular_file() && entry.path().extension() == ".sql";
+                    }) | std::views::transform([](auto const &entry) {
+                        static auto i = 1L;
+                        return std::make_pair(i++, entry);
+                    });
+            for (auto const &[index, migration]: migrations) {
+                migrator.addFile(index, migration.path().string());
+            }
```

Comment:
Since we are globing the entire directory, it'd be nice to see a log with the name of the file which will be used for migration.

Reply:
I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files. What would happen if it'd try to apply the migrations in incorrect order? (e.g 002_sample_data.sql then 001_init.sql)

Reply:
Also this could be simple for loop. Are we bumping the c++ standard just for this?

Reply:
> I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files.

[No, it does not.](https://eel.is/c++draft/filesystems#fs.class.directory.iterator.general-8)

Reply:
> Also this could be simple for loop. Are we bumping the c++ standard just for this?

Boomer detected, opinion rejected

Reply:
> > I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files.
> 
> [No, it does not.](https://eel.is/c++draft/filesystems#fs.class.directory.iterator.general-8)

What should we do  to ensure the order?

Reply:
> > Also this could be simple for loop. Are we bumping the c++ standard just for this?
> 
> Not because of that - you can use all of this in C++20 but I want to use the latest standard if it's not a problem.

If we don't use anything from the latest standard, then it is better to stick with c++20 so we can support more compiler versions.

Reply:
> > > I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files.
> > 
> > 
> > [No, it does not.](https://eel.is/c++draft/filesystems#fs.class.directory.iterator.general-8)
> 
> What should we do to ensure the order?

It think that specifying the migrations by hand would be best.
So just something like this:
```c++
migrator.addFile(1, MAIN_DATABASE_MIGRATIONS_PATH "/001_init.sql");
migrator.addFile(2, MAIN_DATABASE_MIGRATIONS_PATH "/002_sample_data.sql");
```

Reply:
I disagree, migrations are supposed to modify database to new version on software update and there should be more than those two

Reply:
Okay then what will we do about `/002_sample_data.sql`. This migration should only be applied on dev builds when we are testing the app. In production this migration should be omitted. In this case manually specifying the migrations would be better. We could do something like this:

```c++
#ifdef DEV_BUILD
migrator.addFile(i++, MAIN_DATABASE_MIGRATIONS_PATH "/002_sample_data.sql");
#endif
```

If we go with globing the entire directory, we need a way to filter all dev migrations.

Reply:
> I disagree, migrations are supposed to modify database to new version on software update and there should be more than those two

Nothing stops us from adding a new migration on update.

Reply:
> Okay then what will we do about `/002_sample_data.sql`. This migration should only be applied on dev builds when we are testing the app. In production this migration should be omitted. In this case manually specifying the migrations would be better. We could do something like this:
> 
> ```c++
> #ifdef DEV_BUILD
> migrator.addFile(i++, MAIN_DATABASE_MIGRATIONS_PATH "/002_sample_data.sql");
> #endif
> ```
> 
> If we go with globing the entire directory, we need a way to filter all dev migrations.

This migration should be removed tbh and all the files from the migration directory should be applied in ascending order.

Reply:
We could also assume that all files in that directory are SQL files (eg. some may not have `.sql` extension if user adds it) and dispose of views... I just wanted to use it here but not using it should make the things easier to sort (no sort for views pipe syntax even in C++23 atm)

Reply:
resolved in recent changes